### PR TITLE
chore(fw): add docstrings to all opcodes in opcodes enum

### PR DESCRIPTION
**Pull request by ThreeHrSleep**
## 🗒️ Description

Added docstrings to most of the opcodes in the Opcodes Enum.
Left `TSTORE`,`TLOAD`,`SENDALL`,`RJUMPI`,`RJUMPV` incomplete due to confusion
## 🔗 Related Issues
#383 

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [X] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](../docs/) directory.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [X] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

File: src/ethereum_test_tools/vm/opcode.py [hunk 1]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
-    CREATE2 = Opcode(0xF5, popped_stack_items=4, pushed_stack_items=1)
-
-    STATICCALL = Opcode(0xFA, popped_stack_items=6, pushed_stack_items=1)
-
-    REVERT = Opcode(0xFD, popped_stack_items=2)
-    INVALID = Opcode(0xFE)
-
-    SELFDESTRUCT = Opcode(0xFF, popped_stack_items=1)
-    SENDALL = Opcode(0xFF, popped_stack_items=1)
+STOP = Opcode(0x00)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:208]: Indentation is missing in this and all subsequent lines.
```suggestion
    STOP = Opcode(0x00)
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 2]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+Get the amount of available gas, including the corresponding reduction for the cost of this instruction
+
+Inputs
+----
+- None
+
+Outputs
+----
+- Source: [evm.codes/#5A](https://www.evm.codes/#5A)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2033]: Typo.

File: src/ethereum_test_tools/vm/opcode.py [hunk 3]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+Gas
+----
+3
+
+Source: [evm.codes/#81](https://www.evm.codes/#81)
+"""
+
+DUP3 = Opcode(0x82, pushed_stack_items=1, min_stack_height=3)
+"""
+DUP15(a, b, c) = c, a, b, c
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:3175]:
```suggestion
DUP3(a, b, c) = c, a, b, c
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 4]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+Gas
+----
+3
+
+Source: [evm.codes/#82](https://www.evm.codes/#82)
+"""
+
+DUP4 = Opcode(0x83, pushed_stack_items=1, min_stack_height=4)
+"""
+DUP15(value1, value2, value3, value) = value, value2, value3, value
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:3208]:
```suggestion
DUP4(value1, value2, value3, value) = value, value1, value2, value3, value
```
Similar issue on the rest of the DUP opcodes.

File: src/ethereum_test_tools/vm/opcode.py [hunk 5]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+Gas
+----
+3
+
+Source: [evm.codes/#91](https://www.evm.codes/#91)
+"""
+
+SWAP3 = Opcode(0x92, min_stack_height=4)
+"""
+SWAP3(a, value2, value3, b) = b value2, value3, a
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:3880]:
```suggestion
SWAP3(a, value2, value3, b) = b, value2, value3, a
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 6]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+Gas
+----
+3
+
+Source: [evm.codes/#92](https://www.evm.codes/#92)
+"""
+
+SWAP4 = Opcode(0x93, min_stack_height=5)
+"""
+SWAP4(a, value2, value3, value4, b) = b value2, value3, value4, a
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:3914]:
```suggestion
SWAP4(a, value2, value3, value4, b) = b, value2, value3, value4, a
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 7]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+3
+
+Source: [evm.codes/#9F](https://www.evm.codes/#9F)
+"""
+
+
+LOG0 = Opcode(0xA0, popped_stack_items=2)
+"""
+LOG3(offset, size)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:4538]:
```suggestion
LOG0(offset, size) 
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 8]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+- to: Target address
+
+Outputs
+----
+- None
+
+Fork
+----
+Frontier
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5093]:
```suggestion
Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 9]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
-    STATICCALL = Opcode(0xFA, popped_stack_items=6, pushed_stack_items=1)
-
-    REVERT = Opcode(0xFD, popped_stack_items=2)
-    INVALID = Opcode(0xFE)
-
-    SELFDESTRUCT = Opcode(0xFF, popped_stack_items=1)
-    SENDALL = Opcode(0xFF, popped_stack_items=1)
+STOP = Opcode(0x00)
+"""
+STOP()
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:210]:
```suggestion
STOP()
```
Tox (the linting and code format tool we currently use) is complaining about trailing white spaces in some of the lines.

In vscode you could search for `\s+$` to automatically remove all of these.

File: src/ethereum_test_tools/vm/opcode.py [hunk 10]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+"""
+
+TLOAD = Opcode(0x5C, popped_stack_items=1, pushed_stack_items=1)
+"""
+TLOAD(key) = value
+----
+
+Description
+----
+Load a value from the storage at the specified key.
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2081]:
```suggestion
Load a value from the transient storage at the specified key.
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 11]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+- key: The key in storage.
+
+Outputs
+----
+- value: The value associated with the given key.
+
+Fork
+----
+
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2093]:
```suggestion
Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 12]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+"""
+
+TSTORE = Opcode(0x5D, popped_stack_items=2)
+"""
+TSTORE(key, value)
+----
+
+Description
+----
+Store a value in the storage at the specified key.
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2109]:
```suggestion
Store a value in the transient storage at the specified key.
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 13]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+
+Inputs
+----
+- key: The key in storage.
+- value: The value to be stored.
+
+Fork
+----
+
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2118]:
```suggestion
Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 14]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+- src: byte offset in the calldata to copy.
+- length: byte size to copy
+
+Outputs
+----
+- None
+
+Fork
+----
+Frontier
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2148]:
```suggestion
Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 15]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+- None
+
+Outputs
+----
+- value: pushed value, equal to 0
+
+Fork
+----
+Frontier
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2208]:
```suggestion
Shanghai
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 16]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+----
+Jump to a relative position in code.
+
+Inputs
+----
+- destination: The relative position to jump to.
+
+Fork
+----
+Revolution
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:4707]:
```suggestion
N/A
```
It's supposed to be included with EOF, which is aimed for the Prague hard-fork, but until it's formally considered for inclusion we can just put `N/A` here.

File: src/ethereum_test_tools/vm/opcode.py [hunk 17]
```diff
@@ -205,171 +205,4898 @@ class Opcodes(Opcode, Enum):
...
+
+Gas
+----
+375 + 375 * topic_count + 8 * size + memory_expansion_cost
+
+Source: [evm.codes/#A4](https://www.evm.codes/#A4)
+"""
+
+
+RJUMP = Opcode(0xE0, data_portion_length=2)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:4692]: Description for `RJUMP`, `RJUMPI` and `RJUMPV` can be found here: https://notes.ethereum.org/@ipsilon/mega-eof-specification#New-Behavior

**Review by marioevz — COMMENTED** [review]
Looks amazing! Thanks for this. I added some minor comments before we can merge.

**Comment by ThreeHrSleep** [comment]
Thanks for reviewing! 
Implemented the suggested changes.
Please lmk if there's anything that needs to be modified :)

**Review by danceratopz — CHANGES_REQUESTED** [review]
This is looking really great, thanks a lot for adding this! :100: 

In general, `tox` is really helpful to make sure no errors have crept in. The framework-specific checks (`tox -e framework`) and the tests checks (`tox -e tests`) caught the error in the PUSH opcode values (see below). Docs here: https://ethereum.github.io/execution-spec-tests/v2.1.0/writing_tests/verifying_changes/

You will probably come across all sort of whitespace / max-line-length errors when you try to run tox...

These could be either handled in your script, by wrapping the new docstring lines at the relevant column (99) or by using VS Code and plugins. In addition to the repo's VS Code setup you might need a plugin like https://stkb.github.io/Rewrap/ to help with bullet point formatting/line-wrapping.

If you feel it's worth getting VS code setup to handle these, see https://ethereum.github.io/execution-spec-tests/v2.1.0/getting_started/setup_vs_code/.

If not (or if your having other issues), feel free to ping me, and I'm happy to help with formatting to get this over the finish line! :slightly_smiling_face:

**Comment by ThreeHrSleep** [comment]
Thanks a lot for pointing out the problems @danceratopz !! :heart: 
I'll try to get the VS Code setup done,handle the whitespaces & fix the issue with PUSH
will definitely let you know if I get stuck somewhere :)

File: src/ethereum_test_tools/vm/opcode.py [hunk 18]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - value: 32-byte value corresponding to that key. 0 if that key was never written before
+
+    Fork
+    ----
+    Frontier
+
+    Gas
+    ----
+    100
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:1877]:
```suggestion
    warm: 100
    cold: 2100
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 19]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - None
+
+    Fork
+    ----
+    Frontier
+
+    Gas
+    ----
+    100
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:1906]:
```suggestion
    set: 20000
    reset: 2900
```
There's a lot of nuances to the actual gas cost of this opcode, but we can leave this snippet as this for now.

File: src/ethereum_test_tools/vm/opcode.py [hunk 20]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    Get the amount of available gas, including the corresponding reduction for the cost of this
+    instruction
+
+    Inputs
+    ----
+    - None
+
+    Outputs
+    ----
+    - Source: [evm.codes/#5A](https://www.evm.codes/#5A)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2036]:
```suggestion
    - gas_remaining: remaining gas for the current call frame
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 21]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - key: The key in storage.
+
+    Outputs
+    ----
+    - value: The value associated with the given key.
+
+    Fork
+    ----
+
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2096]:
```suggestion
    Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 22]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    Store a value in the storage at the specified key.
+
+    Inputs
+    ----
+    - key: The key in storage.
+    - value: The value to be stored.
+
+    Fork
+    ----
+
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2121]:
```suggestion
    Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 23]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    - src: byte offset in the calldata to copy.
+    - length: byte size to copy
+
+    Outputs
+    ----
+    - None
+
+    Fork
+    ----
+    Frontier
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2151]:
```suggestion
    Cancun
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 24]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - None
+
+    Outputs
+    ----
+    - value: pushed value, equal to 0
+
+    Fork
+    ----
+    Frontier
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:2208]:
```suggestion
    Shanghai
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 25]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    Jump to a relative position in code.
+
+    Inputs
+    ----
+    - destination: The relative position to jump to.
+
+    Fork
+    ----
+    Revolution
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:4737]:
```suggestion
    
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 26]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    None
+
+    Outputs
+    ----
+    None
+
+    Fork
+    ----
+    Homestead
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5079]:
```suggestion
    Frontier
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 27]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    Gas
+    ----
+    5000
+
+    Source: [evm.codes/#FF](https://www.evm.codes/#FF)
+    """
+
     SENDALL = Opcode(0xFF, popped_stack_items=1)
+    """
+    SENDALL(to) = success
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5114]:
```suggestion
    SENDALL(to)
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 28]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+
+    Description
+    ----
+    Send all ether to the target address.
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5119]:
```suggestion
    Send all ether to the target address and halt execution.
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 29]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - None
+
+    Fork
+    ----
+    Frontier
+
+    Gas
+    ----
+
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5135]:
```suggestion
    5000
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 30]
```diff
@@ -198,178 +192,4947 @@ class Opcodes(Opcode, Enum):
...
+
+    Source: [eips.ethereum.org/EIPS/eip-4758](https://eips.ethereum.org/EIPS/eip-4758)
```
Comment (marioevz) [inline src/ethereum_test_tools/vm/opcode.py:5137]:
```suggestion
    Source: [eips.ethereum.org/EIPS/eip-6780](https://eips.ethereum.org/EIPS/eip-6780)
```

**Review by marioevz — CHANGES_REQUESTED** [review]
Some more comments thanks!

I think you should be able to just batch-commit all the comments directly from here.

**Comment by ThreeHrSleep** [comment]
Thanks ! 
and sorry for making so many mistakes in the PR :sweat_smile:

**Review by marioevz — APPROVED** [review]

**Comment by marioevz** [comment]
Hey @danceratopz I think we are unable to merge unless you approve the requested changes, if you'd like to give this another look :+1:

File: src/ethereum_test_tools/vm/opcode.py [hunk 31]
```diff
@@ -198,178 +192,4945 @@ class Opcodes(Opcode, Enum):
...
+    ----
+    - prevRandao: previous block's RANDAO mix
+
+    Fork
+    ----
+    Merge
+
+    Gas
+    ----
+    20
```
Comment (danceratopz) [inline src/ethereum_test_tools/vm/opcode.py:1570]:
```suggestion
    2
```

File: src/ethereum_test_tools/vm/opcode.py [hunk 32]
```diff
@@ -198,178 +192,4945 @@ class Opcodes(Opcode, Enum):
...
+    - None
+
+    Fork
+    ----
+    Frontier
+
+    Gas
+    ----
+    set: 20000
+    reset: 2900
```
Comment (danceratopz) [inline src/ethereum_test_tools/vm/opcode.py:1904]:
```suggestion
    static_gas = 0

    if value == current_value
        if key is warm
            base_dynamic_gas = 100
        else
            base_dynamic_gas = 100
    else if current_value == original_value
        if original_value == 0
            base_dynamic_gas = 20000
        else
            base_dynamic_gas = 2900

    if key is cold:
        base_dynamic_gas += 2100
```

Comment (danceratopz) [reply src/ethereum_test_tools/vm/opcode.py:1904]: Maybe too much :see_no_evil: :laughing:

File: src/ethereum_test_tools/vm/opcode.py [hunk 33]
```diff
@@ -198,178 +192,4945 @@ class Opcodes(Opcode, Enum):
...
+    - offset: byte offset in the memory in bytes. The return data of the calling context.
+    - size: byte size to copy (size of the return data).
+
+    Fork
+    ----
+    Byzantium
+
+    Gas
+    ----
+    0
```
Comment (danceratopz) [inline src/ethereum_test_tools/vm/opcode.py:5053]:
```suggestion
    memory_expansion_cost
```

**Review by danceratopz — CHANGES_REQUESTED** [review]
Sorry for the hold up! Looking good, thank you once again! 

I replaced the unnecessary UTF8 characters in 6864f5d9 (`’` -> `'`). Otherwise, a few fixes/suggestions below!

**Review by danceratopz — APPROVED** [review]

**Comment by danceratopz** [comment]
> Hey @danceratopz I think we are unable to merge unless you approve the requested changes, if you'd like to give this another look 👍

Hey @marioevz, @spencer-tb's doing a last pass :sweat_smile: and will give some more input sooon, but I've approved the requested changes to avoid blocking!

**Comment by spencer-tb** [comment]
I added some additional tweaks in: [035821f](https://github.com/ethereum/execution-spec-tests/pull/424/commits/035821faec0443b191b4bddc5f4c91828217dae1)

- General clean-up, removing `.`'s and newlines etc,
- Fixed forkname mismatches,
- Clean up of swap and push opcodes,
- Gas formatting clean up (the markdown wasn't rendering nicely in vscode),
- Ordered the opcodes,
- Removed SENDALL as we don't use it anywhere,
- Tox fixes.

**Review by spencer-tb — APPROVED** [review]
Cheeky approve after my commit :D

**Comment by danceratopz** [comment]
Amazing. The most commented PR in the history of execution-spec-tests! Thanks again @ThreeHrSleep!

**Pull request merged**